File: diff.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "define.h"
#include "git-root.h"

char* get_diff(const char* file) {
    char* git_root = get_git_root();

    char cmd[512];
    snprintf(cmd, sizeof(cmd), "git diff --cached -- %s/%s", git_root, file);

    FILE* fp = _popen(cmd, "r");
    if (!fp) return NULL;

    char* buffer = malloc(100000);
    buffer[0] = '\0';

    while (fgets(cmd, sizeof(cmd), fp)) {
        strcat(buffer, cmd);
    }

    _pclose(fp);
    return buffer;
}

char** get_staged_files(int* count) {
    FILE* fp = _popen("git diff --cached --name-only", "r");
    if (!fp) return NULL;

    char** files = malloc(128 * sizeof(char*));
    char line[512];
    int i = 0;

    while (fgets(line, sizeof(line), fp)) {
        line[strcspn(line, "\r\n")] = 0;
        files[i] = strdup(line);
        i++;
    }

    _pclose(fp);
    *count = i;
    return files;
}

File: diff.go
package main

import (
	"bufio"
	"bytes"
	"os/exec"
	"strings"
	"sync"
)

var diffBufferPool = sync.Pool{
	New: func() interface{} {
		return new(bytes.Buffer)
	},
}

func GetDiff(file string) (string, error) {
	var builder strings.Builder
	builder.Reset()

	root, err := GetGitRoot()
	if err != nil {
		return "", err
	}

	builder.WriteString(root)
	builder.WriteString("/")
	builder.WriteString(file)

	cmd := exec.Command("git", "diff", "--cached", "--", builder.String())

	buf := diffBufferPool.Get().(*bytes.Buffer)
	buf.Reset()
	defer diffBufferPool.Put(buf)

	cmd.Stdout = buf
	if err := cmd.Run(); err != nil {
		return "", err
	}

	return buf.String(), nil
}

func GetStagedFiles() ([]string, error) {
	cmd := exec.Command("git", "diff", "--cached", "--name-only")

	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return nil, err
	}

	if err := cmd.Start(); err != nil {
		return nil, err
	}

	var files []string
	scanner := bufio.NewScanner(stdout)
	for scanner.Scan() {
		files = append(files, scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	if err := cmd.Wait(); err != nil {
		return nil, err
	}

	return files, nil
}
